File: 2024/python/02_framing_names.py
"""02. Framing Names."""


def create_frame(names: list[str]) -> str:
    """
    Create frame.

    Args:
        names (list[str]): List of names to frame

    Returns:
        str: The framed names
    """
    width = len(max(names, key=len)) + 4
    frame = ["*" * width]

    for name in names:
        frame.append(f"* {name.ljust(width - 4, ' ')} *")

    frame.append("*" * width)

    return "\n".join(frame)


print(create_frame(["midu", "madeval", "educalvolpz"]))
# Expected result:
# ***************
# * midu        *
# * madeval     *
# * educalvolpz *
# ***************

print(create_frame(["midu"]))
# Expected result:
# ********
# * midu *
# ********

print(create_frame(["a", "bb", "ccc"]))
# Expected result:
# *******
# * a   *
# * bb  *
# * ccc *
# *******

print(create_frame(["a", "bb", "ccc", "dddd"]))
# Expected result:
# ********
# * a    *
# * bb   *
# * ccc  *
# * dddd *
# ********

File: 2024/javascript/02-framingNames.js
// 02. Framing Names.

/**
 * @param {string[]} names - Array of names to frame
 * @returns {string} The framed names
 */
function createFrame(names) {
  const width = Math.max(...names.map((name) => name.length)) + 4;
  const frame = ["*".repeat(width)];

  names.forEach((name) => {
    frame.push(`* ${name.padEnd(width - 4, " ")} *`);
  });

  frame.push("*".repeat(width));

  return frame.join("\n");
}

console.log(createFrame(["midu", "madeval", "educalvolpz"]));

// Expected result:
// ***************
// * midu        *
// * madeval     *
// * educalvolpz *
// ***************

console.log(createFrame(["midu"]));

// Expected result:
// ********
// * midu *
// ********

console.log(createFrame(["a", "bb", "ccc"]));

// Expected result:
// *******
// * a   *
// * bb  *
// * ccc *
// *******

console.log(createFrame(["a", "bb", "ccc", "dddd"]));
// ********
// * a    *
// * bb   *
// * ccc  *
// * dddd *
// ********
